### ISSUE TITLE
Regression in jOOQ 3.11 when binding unknown data types

### ISSUE THREAD

**Issue by scadgek** [jOOQ/jOOQ#7370]
### Problem description
There is a table `test_table` in a database with a `created` column of `datetime(3)` data type:
```sql
CREATE TABLE `jooqtest`.`test_table` (
  `created` DATETIME(3) NULL
);
```
The query is created as follows:
```java
InsertQuery<Record> query = using( SQLDialect.MYSQL ).insertQuery( table( name( "test_table" ) ) );
query.addValue( field( name( "test_table", "created" ) ), new java.util.Date() );
```
When executing it directly via `execute`, it correctly transforms the date value and inserts the data row as expected:
```java
using( connection, SQLDialect.MYSQL ).execute( query );
```
But when executing it in batch (with 2 and more queries like this), an exception is thrown:
```java
using( connection, SQLDialect.MYSQL )
            .batch( query, query )
            .execute();
```
Exception stack trace:
```
Exception in thread "main" org.jooq.exception.DataAccessException: SQL [insert into `test_table` (`created`) values ('Mon Apr 02 14:42:59 EEST 2018')]; Data truncation: Incorrect datetime value: 'Mon Apr 02 14:42:59 EEST 2018' for column 'created' at row 1
	at org.jooq_3.10.6.MYSQL.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2241)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:704)
	at org.jooq.impl.BatchMultiple.execute(BatchMultiple.java:123)
	at org.jooq.impl.BatchMultiple.execute(BatchMultiple.java:75)
	at JooqTest.main(JooqTest.java:28)
Caused by: java.sql.BatchUpdateException: Data truncation: Incorrect datetime value: 'Mon Apr 02 14:42:59 EEST 2018' for column 'created' at row 1
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.cj.core.util.Util.handleNewInstance(Util.java:185)
	at com.mysql.cj.core.util.Util.getInstance(Util.java:168)
	at com.mysql.cj.core.util.Util.getInstance(Util.java:175)
	at com.mysql.cj.jdbc.exceptions.SQLError.createBatchUpdateException(SQLError.java:636)
	at com.mysql.cj.jdbc.StatementImpl.executeBatchInternal(StatementImpl.java:1104)
	at com.mysql.cj.jdbc.StatementImpl.executeBatch(StatementImpl.java:1009)
	at org.jooq.tools.jdbc.DefaultStatement.executeBatch(DefaultStatement.java:100)
	at org.jooq.tools.jdbc.DefaultStatement.executeBatch(DefaultStatement.java:100)
	at org.jooq.impl.BatchMultiple.execute(BatchMultiple.java:104)
	... 2 more
Caused by: com.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: Incorrect datetime value: 'Mon Apr 02 14:42:59 EEST 2018' for column 'created' at row 1
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:97)
	at com.mysql.cj.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:1983)
	at com.mysql.cj.jdbc.StatementImpl.executeUpdateInternal(StatementImpl.java:1603)
	at com.mysql.cj.jdbc.StatementImpl.executeBatchInternal(StatementImpl.java:1076)
	... 6 more
```

### Additional findings
When only a single query of this type is executed in batch, the same exception is thrown, but also a log message is displayed before throwing an exception:
```java
using( connection, SQLDialect.MYSQL )
            .batch( query )
            .execute();
```
```
INFO: Single batch             : No bind variables have been provided with a single statement batch execution. This may be due to accidental API misuse
```
Which gave me a clue how to fix it by adding a `bind` call:
```java
using( connection, SQLDialect.MYSQL )
            .batch( query )
            .bind( new java.util.Date() )
            .execute();
```
This way, the exception wasn't thrown and the data row was saved successfully in a database. I would happily go ahead with this workaround, but my SQL inserts aren't of the same form, e.g. one can be `insert into table (column1, column2) (?,?)` and the other can be `insert into table (column2, column3, column4) (?,?,?)` and these should be executed in batch - so obviously I cannot just call `bind` multiple times since the columns differ.

And finally, [the documentation](https://www.jooq.org/doc/3.10/manual/sql-execution/batch-execution/) states that these two approaches (`batch( query1, query2 )` and `batch( query ).bind( values1 ).bind( values2 )` should have the same behaviour, while in reality they differ in the way they prepare values for the SQL statement - this is why I believe this is a bug.

### Versions:

- jOOQ: 3.10.6
- Java: 8
- Database (include vendor): MySQL 5.7.20
- OS: Linux Mint 
- JDBC Driver (include name if inofficial driver): [mysql-connector-java 6.0.6](http://mvnrepository.com/artifact/mysql/mysql-connector-java/6.0.6)

**Comment by lukaseder** [comment]
Interesting, thanks for your report. The main problem here is your usage of `java.util.Date`, which isn't the most appropriate type corresponding to SQL `TIMESTAMP` or `DATETIME` types. Better use `java.sql.Timestamp` or `java.time.LocalDateTime`, instead. `java.util.Date` is poorly supported by jOOQ.

However, since it works with non-batch usage, we should probably make it work with batch usage as well.

**Comment by scadgek** [comment]
Indeed, my samples above do work with `LocalDateTime.now()` instead of `new java.util.Date()`. Unfortunately, in the original project I cannot currently use it due to some legacy reasons, but probably I'll try to at move in this direction at least.

Anyway, either fixing this for `java.util.Date` or giving an appropriate warning or error would be beneficial.

Thank you for the assistance!

**Comment by lukaseder** [comment]
> Unfortunately, in the original project I cannot currently use it due to some legacy reasons, but probably I'll try to at move in this direction at least.

Can't you use `java.sql.Timestamp` either?

**Comment by lukaseder** [comment]
... another option is to register a `Converter` or `Binding` with your column:

```java
DataType<java.util.Date> dataType = SQLDataType.TIMESTAMP.asConvertedDataType(new MyConverter);
field( name( "test_table", "created" ), dataType )
```

**Comment by lukaseder** [comment]
In jOOQ 3.11, your first working example no longer works. Unsurprising as we've refactored some data type binding internals, and `java.util.Date` really isn't a supported data type. We'll throw a more meaningful `DataTypeException` instead

**Comment by scadgek** [comment]
I will definitely try again both with `LocalDateTime` and with `Timestamp`. Most probably it will be possible to use them instead after a bit of refactoring. Otherwise I can use a converter - thanks for the hint.
By the way, can I use converter for [retrieving `TINYINT(1)` as boolean](https://groups.google.com/d/msg/jooq-user/kW_j3GL7tSI/SJy6e8dWAwAJ) yet?

**Comment by lukaseder** [comment]
> By the way, can I use converter for retrieving TINYINT(1) as boolean yet?

Of course. Just use `field(name("x", "y"), SQLDataType.BOOLEAN)`. If you have any questions about this, please create a new issue as it isn't related.

**Comment by lukaseder** [comment]
The reason why the non-batch insert worked thus far is because apparently, the MySQL JDBC driver supports the `java.util.Date` type when used with `PreparedStatement.setObject()` and jOOQ just routes unknown data types to the JDBC driver this way (which is important for many vendor-specific types).

However, in the `batch()` call, since you're using multi-statement batching, which doesn't support bind variables on the JDBC level, jOOQ inlines your bind variable. The best way to do that is by calling `toString()` on unknown data types. This is why you get different behaviour here.

So, I don't think this is a bug, nor there needs to be anything done about it. This works as designed.

**Comment by lukaseder** [comment]
(Nevertheless, I'll definitely fix the regression in jOOQ 3.11)
